src/shell/shell_misc.cpp

```diff
@@ -20,479 +20,421 @@
-						move_cursor_back_one();
-						outc(' ');
-						move_cursor_back_one();
-					}
-
-					strcpy(&line[completion_index], it_completion->c_str());
-					len = (uint16_t)it_completion->length();
-					str_len = str_index = completion_index + len;
-					size = CMD_MAXLINE - str_index - 2;
-					DOS_WriteFile(STDOUT, (uint8_t *)it_completion->c_str(), &len);
+				auto delimiter = command.find_last_of(" \\");
+				if (delimiter == std::string::npos) {
+					completion_start = 0;
+				} else {
+					completion_start = delimiter + 1;
 				}
 			}
+
+			command.erase(completion_start);
+			command += completion[completion_index];
+			cursor_position         = command.size();
+			viewing_tab_completions = true;
 			break;
-		case 0x1b: /* Esc */
-			while (str_index < str_len) {
-				outc(line[str_index++]);
+		}
+
+		case '\b':
+			if (cursor_position > 0) {
+				command.erase(cursor_position - 1, 1);
+				cursor_position--;
 			}
-			//write a backslash and return to the next line
-			outc('\\');
-			outc('\r');
-			outc('\n');
-			reset_str(line);
-			if (l_completion.size()) l_completion.clear(); //reset the completion list.
-			this->InputCommand(line);	//Get the NEW line.
-			size = 0;       // stop the next loop
-			str_len = 0;    // prevent multiple adds of the same line
 			break;
+
+		case '\n': break;
+		case '\r': int10_manager.Newline(); return command;
+
+		case Escape:
+			command += "\\";
+			int10_manager.Update(command, cursor_position);
+			int10_manager.Newline();
+			command.clear();
+			cursor_position = 0;
+			break;
+
 		default:
-			if (l_completion.size()) l_completion.clear();
-			if(str_index < str_len && true) { //mem_readb(BIOS_KEYBOARD_FLAGS1)&0x80) dev_con.h ?
-				outc(' ');//move cursor one to the right.
-				auto text_len = static_cast<uint16_t>(str_len - str_index);
-				uint8_t* text=reinterpret_cast<uint8_t*>(&line[str_index]);
-				DOS_WriteFile(STDOUT, text, &text_len); // write buffer to screen
-				move_cursor_back_one(); //undo the cursor the right.
-				for (auto i = str_len; i > str_index; i--) {
-					line[i]=line[i-1]; //move internal buffer
-					move_cursor_back_one(); //move cursor back (from write buffer to screen)
-				}
-				// new end (as the internal buffer moved one
-				// place to the right
-				terminate_str_at(line, ++str_len);
-				size--;
-			};
-
-			line[str_index]=c;
-			str_index ++;
-			if (str_index > str_len){ 
-				terminate_str_at(line, str_index);
-				str_len++;
-				size--;
+			if (Int10Manager::MaxCommandSize() < command.size()) {
+				break;
 			}
-			DOS_WriteFile(STDOUT,&c,&n);
+			command.insert(cursor_position, 1, static_cast<char>(data));
+			cursor_position++;
 			break;
 		}
+
+		int10_manager.Update(command, cursor_position);
+
+		if (!viewing_tab_completions) {
+			completion.clear();
+		}
 	}
 
-	if (!str_len) return;
-	str_len++;
+	return "";
+}
+
+namespace {
+std::vector<std::string> get_completions(const std::string_view command)
```

Comment:
We're not using this nameless `namespace` style in other refactored code.  Maybe `static` is sufficient to scope the function locally?

Reply:
I'm totally neutral and just sticking to the existing pattern ; so unless we want to make a mass-switch, might as well carry on w/ `static` until there's a big improvement to be had.

Once C++20 or 23 gives us modules, we'll likely make whatever changes are needed en-mass to get the best bang for the buck.